Reporting fix and touchup

jmartin-sul opened the pull request:
## Why was this change made?

(follow on to #1360)

initially i started in on a minor bit of refactoring so that more could be done in the initial query (which was previously just used to list druids), since that query already collects almost all the data we might want to report.  when a test failed i went to do a bit of manual testing, and noticed that the rake tasks weren't actually working, so i fixed those in addition to getting rid of the need to collect the druid list separately from the loop that collates results.  then i noticed that the reports all used the same default file name no matter the type, so i fixed that by adding a report type param to the csv writing method.

i also finally realized we probably want column headers (after waffling on that when first creating the `Reporter` class), so i added those.

did some refactoring and test fixing along the way to support the above.  more detail in the commit messages (the first one's a bit of a grab bag, but i tried to make the other ones more focused).

in the last commit i finally did the last bit of the query optimization i'd originally set out to do, using the `complete_moabs` info that the main query already provided, and doing away with the need to instantiate lots of ActiveRecord objects.

two other things i'd like to do, but which i deferred because it's late:
1. rename the class to `MoabStorageRootReporter` because it's now very moab specific (i probably should've named it that from the beginning).
2. turn the `Reporter` methods that don't take advantage of internal state back into class methods (open to argument on this -- i get the impression some prefer to avoid class methods entirely? i don't have strong feelings on this, but my inclination is to not make something an instance method if it doesn't need to be)

## Was the usage documentation (e.g. wiki, README, queue or DB specific README) updated?

n/a

justinlittman commented:
If I'm reading correctly, these reports read all of the data into memory before writing to CSV. I don't know the size of the reports. Is this problematic?

mjgiarlo commented on app/services/reporter.rb:
```diff
@@ -6,61 +6,80 @@
-  # @return [Array] an array of druids on the storage root
-  def moab_storage_root_druid_list
-    druid_array = []
+  def moab_storage_root_list_pres_obj_comp_moabs_relation
```
I appreciate renaming this method to be more descriptive of what it's doing. If I'm grokking (and I may not be), this method is primarily operating on PreservedObjects, using CompleteMoabs only to filter the POs in the given storage root. If that is true, I wonder about a tweak to the name like `preserved_objects_in_storage_root`? (The name currently is almost Germanic in its love for compound nouns, which is impressive. :smile:)

mjgiarlo replied:
And I might also like to see some YARD documentation to show that it returns an AR relation. (Would rather see that in YARD than in the method name, but I'm open to being overruled there! For me, I think having `preserved_objects` in the method name already implies it's an AR relation but maybe that's just me.)

jmartin-sul replied:
> The name currently is almost Germanic in its love for compound nouns, which is impressive

ha, glad you like it, i think my tendency for that drives @ndushay crazy 😆 

> this method is primarily operating on PreservedObjects, using CompleteMoabs only to filter the POs in the given storage root

sorta.  i found it more useful to think of this in SQL than in ActiveRecord objects.

this method is really about getting all the `CompleteMoabs` on a storage root, and all their relevant details.  `PreservedObject` is the master record for an object in the abstract, tying together its uncompressed moabs on storage roots and the zip copies in the cloud (and the preservation policies that say what goes where).  as such, the `druid` lives on the `PreservedObject`.  that's the main point of tying in the `PreservedObject`.

so really i want all the info for `CompleteMoab`s on the storage root, plus some info that lives on the related `PreservedObject` and `MoabStorageRoot` rows.  might've been better to have the query be rooted in `PreservedObject`, joining to `CompleteMoab`?  i don't have any strong feelings on that, esp since all i really want to do is get the query results as a list of column values, without bothering to instantiate ActiveRecord objects.

i think this is just a natural artifact of the tendency for OOP to clash with relational modeling at the seams where the two meet.  since this is a query for a large number of things that we want to run correctly and efficiently, i erred on the side of writing good SQL (in the most maintainable rails way i could imagine) rather than erring on the side of writing things in the rails-iest way.  but i also wanted to use AR relation composability instead of just writing straight SQL or something.  very open to feedback on how to check all those boxes in a cleaner way.

jmartin-sul replied:
> And I might also like to see some YARD documentation to show that it returns an AR relation. (Would rather see that in YARD than in the method name, but I'm open to being overruled there! For me, I think having preserved_objects in the method name already implies it's an AR relation but maybe that's just me.)

works for me.  will try to get to a terser name in light of YARD doc for return type.  if you have suggestions on naming in light of the data i'm trying to get back out of the relation, i'm all ears.

mjgiarlo replied:
Aha! Now that I look at how this relation is used and cross-reference it with the database schema, I **now** see that the PreservedObject is primarily for the druid, and that all of the other fields come from the CompleteMoab. How about a name like `complete_moabs_and_preserved_objects_in_storage_root` or some such? This is me saying that I am OK with verbose method names and that I also prefer in-domain language over abbreviations, which is :100: subjective.

ndushay replied:
I'm thinking this method should be private.  Does that ring true for you?

mjgiarlo replied:
@ndushay I think that makes good sense.

mjgiarlo replied:
@ndushay adding to that, I think only these methods should be public:

* `#initialize`
* `#write_to_csv`
* `#druid_csv_list`
* `#moab_detail_csv_list`

The rest could/should be private to constrain the interface of the reporter and hide its implementation details.

jmartin-sul replied:
> How about a name like `complete_moabs_and_preserved_objects_in_storage_root` or some such?

i like it, will go w/ that

jmartin-sul replied:
i took all the suggestions for privatization, except for `#default_filepath`:  it was easier to leave that public in this follow up because it kept me from having to change the tests at all, and it seems defensible to leave it in the public interface (but that's me rationalizing convenience, and i'm happy to privatize that too, but would prefer to do so in a follow-on PR).

mjgiarlo replied:
I'm not seeing the changes yet, @jmartin-sul. did you push?

jmartin-sul replied:
`#default_filepath` will be made private as part of addressing #1367

mjgiarlo commented on app/services/reporter.rb:
```diff
@@ -6,61 +6,80 @@
+    druid_array = [['druid']]
+    moab_storage_root_list_pres_obj_comp_moabs_relation
+      .select(:druid)
       .each_row do |po_hash|
```
I'm not that familiar with the `#each_row` implementation. Will that loop over rows in the DB without loading it all into memory? Maybe this is also what @justinlittman was asking.

jmartin-sul replied:
`#each_row` is from this gem:  https://github.com/afair/postgresql_cursor

it allows the result set to be returned in pages using a real database cursor, without the jankiness of `find_each`.  in the #1347 implementation, it _was_ used to address the concern of loading everything into memory before writing, but that was lost in the DRYing of #1360.  i could go either way on whether we should spend effort on getting back to writing the CSV out within the query paging instead of loading the full result set into memory.

mjgiarlo replied:
I am totally fine with going forward with this and optimizing for performance/robustness later on if it is determined that this is *actually* problematic.

mjgiarlo commented on app/services/reporter.rb:
```diff
@@ -6,61 +6,80 @@
+  def druid_csv_list
+    druid_array = [['druid']]
+    moab_storage_root_list_pres_obj_comp_moabs_relation
+      .select(:druid)
```
Is there any chance we will get duplicate druids in this list? And, if we do, do we care about druids being in the CSV list multiple times?

jmartin-sul replied:
it is possible in that the schema allows for multiple `CompleteMoab`s for a given `PreservedObject`.  in practice, we never got around to supporting this in the rest of the code, and some of our shameless green implementations elsewhere assume one `CompleteMoab` per `PreservedObject`, iirc.  also, in practice, i don't think we'd put the same moab on the same storage root multiple times.  so i think that in practice, the druids will be unique in this list, both because druids are currently unique across `CompleteMoab`s, and because even in our more robust future, they'd likely still be unique for all `CompleteMoab`s on a given storage root.

mjgiarlo replied:
As long as you think this is unlikely to cause heartache or confusion, this is all good. Otherwise, we could do a distinct query but that would open up the can of worms of which row to  prefer over the others.

ndushay replied:
we could add a comment stating `there will only be uniq druid values for a single storage root`

mjgiarlo commented on app/services/reporter.rb:
```diff
@@ -6,61 +6,80 @@
-                          storage_root: cm.moab_storage_root.name,
-                          from_storage_root: cm.from_moab_storage_root&.name }
+  def moab_detail_csv_list(errors_only: false)
+    query =
```
Dig this change, and I'm happy to see relations passed around for chaining just like this.

jmartin-sul replied:
this composability is literally the thing that softened my general dislike for heavy/magical ORMs like ActiveRecord 🙂

mjgiarlo commented on app/services/reporter.rb:
```diff
@@ -6,61 +6,80 @@
+        moab_storage_root_list_pres_obj_comp_moabs_relation
       end
+
+    header_row = [['druid', 'from_storage_root', 'storage_root', 'last_checksum_validation', 'last_moab_validation', 'status', 'status_details']]
```
This might could be a constant or its own method or some such, but I don't think it's too worth fussing over.

jmartin-sul replied:
oh yeah, hadn't thought of that!  happy to do that if you'd prefer.  i also sorta like having it right by the row that's being constructed, but i don't feel strongly about that (and that wasn't intentional, that's just how i ended up writing it).

mjgiarlo replied:
Do not feel strongly about this.

ndushay replied:
"from_storage_root" sucks.  How about "previous storage root"?   And column headers wouldn't need the underscores -- just make's 'em harder to read.

ndushay replied:
I started my own branch with some of these fixes, and I just added the header row directly without assigning it a variable name.  I don't have strong feelings.  <feelings ... nothing more than feelings ...>

mjgiarlo commented on app/services/reporter.rb:
```diff
@@ -6,61 +6,80 @@
 
-  # @param [Array] druids - list of druids to output details for
   # @param [Boolean] errors_only (default: false) - optionally only output lines with audit errors
   # @return [Array] an array of hashes with details for each druid provided
```
Is this now returning an array of arrays? How about `# @return [Array<Array>] an array of arrays, with each inner array representing a CSV row (including a header row)` ?

jmartin-sul replied:
ah, yes!  good catch.

mjgiarlo commented on app/services/reporter.rb:
```diff
@@ -6,61 +6,80 @@
+  end
+
+  def moab_storage_root_name(msr_id)
+    return nil unless msr_id.present?
```
In this case, I have a slight preference for `return nil if msr_id.blank?` for being just a little easier (for me?) to grok. But it's no biggie.

jmartin-sul replied:
fin by me, will do

mjgiarlo commented on app/services/reporter.rb:
```diff
@@ -6,61 +6,80 @@
+
+  def status_text_from_code(status_code)
+    # statuses is a hash of the form { 'status_text' => status_code }.  this is just a reverse hash lookup.
+    CompleteMoab.statuses.to_a.find { |s| s[1] == status_code }[0]
```
I think you can shorten this to `CompleteMoab.statuses.key(status_code)`.

jmartin-sul replied:
ah, nice, this felt circuitous when i was writing it, will fix

mjgiarlo commented on app/services/reporter.rb:
```diff
@@ -6,61 +6,80 @@
   end
 
   # @param [Array] lines - values to output on each line of the csv
   # @param [String] filename - optional filename to override the default
```
Pls add `@param` documentation of the new `report_type` kwarg.

jmartin-sul replied:
will do

mjgiarlo commented on app/services/reporter.rb:
```diff
@@ -6,61 +6,80 @@
 
   # @param [Array] lines - values to output on each line of the csv
   # @param [String] filename - optional filename to override the default
   # @return [String] the name of the CSV file to which the list was written
```
Pls add `@raise` documentation for the new `ArgumentError` condition.

mjgiarlo replied:
Would you also consider adding `@raise` doc for the `RuntimeError` on line 78?

jmartin-sul replied:
will do

mjgiarlo commented on app/services/reporter.rb:
```diff
@@ -6,61 +6,80 @@
-  def write_to_csv(lines, filename: nil)
-    filename ||= default_filename(filename_prefix: "MoabStorageRoot_#{storage_root.name}_druids", filename_suffix: 'csv')
+  def write_to_csv(lines, report_type: nil, filename: nil)
+    raise ArgumentError, 'Must specify at least one of report_type or filename' unless report_type.present? || filename.present?
```
As above, I might have a slight preference for `if report_type.blank? && filename.blank?`

jmartin-sul replied:
works for me

jmartin-sul commented:
@justinlittman said:
> If I'm reading correctly, these reports read all of the data into memory before writing to CSV. I don't know the size of the reports. Is this problematic?

that's correct.  that wasn't the case before the refactoring in #1360, but the DRYing there made it so that everything was collected first and then passed to the CSV writer method.

i don't personally have any objection to that at this point -- that was how i'd initially approached this reporting work, though i did end up writing from within the query result paging in the reporting PR that initially got merged (#1347).  i'm happy to move back to something that pages through from within in the cursor iteration loop.  fwiw, i think we'd be getting at most a few hundred thousand rows with a few hundred characters of data each.  that feels probably fine to me, esp for a manually triggered report?

either way, i think what's in this PR is an improvement over what's in master at the moment, and doesn't change the loading everything into memory aspect of things either way.  i'd be happy to re-visit not loading everything into memory, but at this point, i think i'd prefer to do that in a separate PR.  i think i have an idea of how i'd do that, but i think i'd want to pair with someone to do it i a readable and idiomatically rails-y way.  my first thought along those lines is something like

jmartin-sul commented:
@mjgiarlo good catch on the YARD stuff.  i realized i missed those updates (and adding YARD docs for the new methods in this PR) after i woke up this morning.  will fix up and ping you once that's done.

jmartin-sul commented:
oh also, sample rake task output:

```
$ be rake prescat:reports:msr_druids['fixture_sr1']
druids for fixture_sr1 written to /preservation_catalog/log/reports/MoabStorageRoot_fixture_sr1_druids_2020-02-12T09:45:09Z.csv
$
$ cat "/preservation_catalog/log/reports/MoabStorageRoot_fixture_sr1_druids_2020-02-12T09:45:09Z.csv"
druid
bj102hs9687
bz514sm9647
jj925bx9565
```

```
$ be rake prescat:reports:msr_druid_detail['fixture_sr1']
druid details for fixture_sr1 written to /preservation_catalog/log/reports/MoabStorageRoot_fixture_sr1_druid_detail_2020-02-12T09:46:34Z.csv
$
$ cat "/preservation_catalog/log/reports/MoabStorageRoot_fixture_sr1_druid_detail_2020-02-12T09:46:34Z.csv"
druid,from_storage_root,storage_root,last_checksum_validation,last_moab_validation,status,status_details
bj102hs9687,,fixture_sr1,2020-02-12 05:29:37 UTC,2020-02-12 05:29:36 UTC,invalid_checksum,validate_checksums (actual location: fixture_sr1; ) Moab file spec/fixtures/storage_root01/sdr2objects/bj/102/hs/9687/bj102hs9687/v0001/data/metadata/versionMetadata.xml was not found in Moab signature catalog spec/fixtures/storage_root01/sdr2objects/bj/102/hs/9687/bj102hs9687/v0003/manifests/signatureCatalog.xml && CompleteMoab status changed from validity_unknown to invalid_checksum
bz514sm9647,,fixture_sr1,2020-02-12 05:29:37 UTC,2020-02-12 05:29:37 UTC,ok,validate_checksums (actual location: fixture_sr1; ) checksum(s) match && CompleteMoab status changed from validity_unknown to ok
jj925bx9565,,fixture_sr1,2020-02-12 05:29:37 UTC,2020-02-12 05:29:37 UTC,ok,validate_checksums (actual location: fixture_sr1; ) checksum(s) match && CompleteMoab status changed from validity_unknown to ok
```

```
$ be rake prescat:reports:msr_audit_errors['fixture_sr1']
druids with errors details for fixture_sr1 written to /preservation_catalog/log/reports/MoabStorageRoot_fixture_sr1_audit_errors_2020-02-12T09:47:12Z.csv
$
$ cat "/preservation_catalog/log/reports/MoabStorageRoot_fixture_sr1_audit_errors_2020-02-12T09:47:12Z.csv"
druid,from_storage_root,storage_root,last_checksum_validation,last_moab_validation,status,status_details
bj102hs9687,,fixture_sr1,2020-02-12 05:29:37 UTC,2020-02-12 05:29:36 UTC,invalid_checksum,validate_checksums (actual location: fixture_sr1; ) Moab file spec/fixtures/storage_root01/sdr2objects/bj/102/hs/9687/bj102hs9687/v0001/data/metadata/versionMetadata.xml was not found in Moab signature catalog spec/fixtures/storage_root01/sdr2objects/bj/102/hs/9687/bj102hs9687/v0003/manifests/signatureCatalog.xml && CompleteMoab status changed from validity_unknown to invalid_checksum
```

ndushay commented on app/services/reporter.rb:
```diff
@@ -6,61 +6,80 @@
+      .select(:druid)
       .each_row do |po_hash|
-        druid_array << po_hash['druid']
+        druid_array << [po_hash['druid']]
```
add comment  ```# csv needs elements to be arrays``` ?  I didn't find that intuitive, and apparently neither did Aaron.

jmartin-sul replied:
did this by adding YARD doc for return type of `#*_csv_list` methods and param type expected by `#write_to_csv`

ndushay commented on app/services/reporter.rb:
```diff
@@ -6,61 +6,80 @@
+        po_cm_hash['druid'], moab_storage_root_name(po_cm_hash['from_moab_storage_root_id']), storage_root.name,
+        po_cm_hash['last_checksum_validation'], po_cm_hash['last_moab_validation'], status_text_from_code(po_cm_hash['status_code']),
+        po_cm_hash['status_details']
+      ]
```
if this won't all fit on one line, my preference is to do each one as its own line.  Possibly same for header row for easy visual analogy.

jmartin-sul replied:
will do this in a follow on PR, if that's ok (sorry, just saw this when combing through to see whether i'd missed any comments -- turns out i did)

jmartin-sul replied:
addressed in fix for #1367

ndushay commented on app/services/reporter.rb:
```diff
@@ -6,61 +6,80 @@
     ensure_containing_dir(filename)
     CSV.open(filename, 'w') do |csv|
       lines.each do |line|
-        line = line.values if line.is_a? Hash
```
yes, nice to get rid of this line;  it was confusing.

ndushay commented on app/services/reporter.rb:
```diff
@@ -6,61 +6,80 @@
-                          last_checksum_validation: cm.last_checksum_validation,
-                          storage_root: cm.moab_storage_root.name,
-                          from_storage_root: cm.from_moab_storage_root&.name }
+  def moab_detail_csv_list(errors_only: false)
```
i'm not a huge fan of named params (params hash - whatever it's called) when there are a small number and the method is unlikely to be used outside the single coded caller ...

mjgiarlo replied:
@ndushay :laughing: welllllll, I'm a huge fan of preferring keyword arguments over standard arguments for two reasons:

1. They are not ordered so a dependency is removed: needing to know what order the args are passed in. We still do have the dependency of needing to know what args are required, but that's true of both styles);
1. They are more intention-revealing than standard arguments. How many times have you seen code like `do_the_thing(42, true)` and wondered WTH `42` and `true` mean? So then you have to go trawling in another codebase, potentially, for enlightenment. Compare with `do_the_thing(max_records: 42, skip_accessioned: true)`. This also means we need to be vigilant about coming up with good names, avoiding names like `process` and `data` and such, but that (again) is true in both styles.

jmartin-sul replied:
i also prefer keyword args for the reasons @mjgiarlo gives.

ndushay commented on app/services/reporter.rb:
```diff
@@ -6,61 +6,80 @@
   # @return [String] the name of the CSV file to which the list was written
-  def write_to_csv(lines, filename: nil)
-    filename ||= default_filename(filename_prefix: "MoabStorageRoot_#{storage_root.name}_druids", filename_suffix: 'csv')
+  def write_to_csv(lines, report_type: nil, filename: nil)
```
same comment about params hash;  I think positional params are fine for the context of this method.

jmartin-sul commented:
@mjgiarlo and @ndushay, i believe i've addressed all of your comments, except for the one i missed from naomi about multi-line lists.

if y'all don't mind, i'll deal with that and the class renaming in a follow on PR.

ndushay commented:
@jmartin-sul i think you can remove "filename_suffix" from the args passed to default_filename, because we don't ever input one;  it's always ```.csv```.   Another thing for the follow on PR.

jmartin-sul commented:
> i think you can remove "filename_suffix" from the args passed to `default_filename`

will be fixed as part of addressing #1367
